sale_payment_sheet/models/sale_payment_sheet.py

```diff
@@ -0,0 +1,336 @@
+    ref = fields.Char(string="Reference")
+    note = fields.Text(string="Notes")
+    sequence = fields.Integer(
+        index=True,
+        help="Gives the sequence order when displaying a list of payment sheet lines.",
+        default=1,
+    )
+    company_id = fields.Many2one(
+        "res.company",
+        related="sheet_id.company_id",
+        string="Company",
+        store=True,
+    )
+    state = fields.Selection(related="sheet_id.state", string="Status")
+    invoice_id = fields.Many2one(
+        comodel_name="account.move", string="Invoice", index=True
+    )
+    transaction_type = fields.Selection(
+        [("partial", "Partial payment"), ("full", "Full payment")],
+        compute="_compute_transaction_type",
+    )
+
+    @api.depends("amount", "invoice_id")
+    def _compute_transaction_type(self):
+        for line in self:
+            amount = (
+                line.amount
+                if line.invoice_id.move_type == "out_invoice"
+                else -line.amount
+            )
+            if float_compare(
+                amount,
+                line.invoice_id.amount_total,
+                precision_digits=line.sheet_id.currency_id.decimal_places,
+            ):
+                line.transaction_type = "partial"
+            else:
+                line.transaction_type = "full"
+
+    @api.depends("sheet_id.user_id", "invoice_id", "transaction_type")
+    def _compute_name(self):
+        for line in self:
+            if not line.create_date:
+                line.create_date = fields.Datetime.now()
+            line.name = "[{}] - {} - {} - ({})".format(
+                fields.Datetime.context_timestamp(line, line.create_date).strftime(
+                    "%H:%M"
+                ),
+                line.sheet_id.user_id.name,
+                line.invoice_id.name,
+                dict(
+                    line._fields["transaction_type"]._description_selection(line.env)
+                ).get(line.transaction_type),
+            )
+
+    @api.depends("invoice_id")
+    def _compute_amount(self):
+        for line in self:
+            amount = line.invoice_id.amount_residual
+            line.amount = (
+                amount if line.invoice_id.move_type == "out_invoice" else -amount
+            )
+
+    def _inverse_amount(self):
+        for line in self:
+            if line.invoice_id.move_type == "out_refund" and line.amount > 0.0:
+                line.amount = -line.amount
+
+    @api.constrains("invoice_id", "amount")
+    def _check_invoice(self):
+        for line in self:
+            # Allow to enter sheet line with an amount of 0,
+            if line.journal_currency_id.is_zero(line.amount):
+                raise ValidationError(
+                    _("The amount of a cash transaction cannot be 0.")
+                )
+            # Do not allow to enter a invoice totally payed more than one time
+            payment_lines = self.search(
```

Comment:
This search should be set before the loop.

Reply:
But this search is necessary inside and not outside because filters by sheet and invoice_id and needs the line in the loop

Reply:
```python
all_payment_lines = self.search([("invoice_id", "in", self.invoice_id.ids), ("sheet_id.state", "=", "open")])
for line in self:
    payment_lines = all_payment_lines.filtered(lambda payment_line, line=line: line.invoice_id == payment_line.id)
    (...)
```

Reply:
Each search generates an SQL query. Doing the search before the loop will be always better.

Reply:
Done, thanks!